service/tablet_allocator.cc

```diff
@@ -25,6 +25,80 @@ namespace service {
 
 seastar::logger lblogger("load_balancer");
 
+struct load_balancer_dc_stats {
+    uint64_t calls = 0;
+    uint64_t migrations_produced = 0;
+    uint64_t migrations_skipped = 0;
+    uint64_t tablets_skipped_node = 0;
+    uint64_t tablets_skipped_rack = 0;
+    uint64_t stop_balance = 0;
+    uint64_t stop_load_inversion = 0;
+    uint64_t stop_no_candidates = 0;
+    uint64_t stop_skip_limit = 0;
+    uint64_t stop_batch_size = 0;
+};
+
+struct load_balancer_node_stats {
+    double load = 0;
+};
+
+using dc_name = sstring;
+
+class load_balancer_stats_manager {
+    std::unordered_map<dc_name, std::unique_ptr<load_balancer_dc_stats>> _dc_stats;
+    std::unordered_map<host_id, std::unique_ptr<load_balancer_node_stats>> _node_stats;
+    seastar::metrics::label dc_label{"target_dc"};
+    seastar::metrics::label node_label{"target_node"};
+    seastar::metrics::metric_groups _metrics;
+
+    void setup_metrics(const dc_name& dc, load_balancer_dc_stats& stats) {
+        namespace sm = seastar::metrics;
+        auto dc_lb = dc_label(dc);
+        _metrics.add_group("load_balancer", {
+            sm::make_counter("calls", sm::description("number of calls to the load balancer"),
+                             stats.calls)(dc_lb),
+            sm::make_counter("migrations_produced", sm::description("number of migrations produced by the load balancer"),
+                             stats.migrations_produced)(dc_lb),
+            sm::make_counter("migrations_skipped", sm::description("number of migrations skipped by the load balancer due to load limits"),
+                             stats.migrations_skipped)(dc_lb),
+        });
+    }
+
+    void setup_metrics(const dc_name& dc, host_id node, load_balancer_node_stats& stats) {
```

Comment:
These setup_metrics() should have different names, they're doing different things (though related).

Reply:
I guess it's too minor, if no other comments I'll merge.